### DISCUSSION TITLE
Feature: "Secret Santa" mode

### DISCUSSION THREAD

**Discussion by cmintey** [cmintey/wishlist#143] [Ideas]
Add a mode where you can randomly select names from and partner two users together, similar to "secret santa". In this mode, a user would only see two lists: theirs, and their partners.

**Comment by trinitrion** [comment]
+1

**Comment by sdgathman** [comment]
I did a cli implementation of this in python.  Very straightforward - the trickiest part was people who shouldn't get each other for secret santa gift.  So I am sharing the simple config used as data, to inspire the design for your version.
```
[santa]
message=Hi %s, Elly told Santa to draw names.  Buy a gift for %s less than $25.  Confirm with Elly.

[givers]
kent: Kent
mary: Mary Beth
susan: Susanna
paul: Paul
elly: Elly
mike: Michael
esther: Esther
dave: David

[block]
susan: paul
paul: susan
mary: kent
kent: mary
esther: dave
dave: esther
```
Note that blocks usually (but not always) go both ways, so having a separate config/option for one way blocks would halve the size of the block list.  Note that drawing yourself is always excluded.

There is a separate database with SMS and email of users - which wishlist already has, so I didn't include it.
Just in case the basic algorithm is not obvious, here is the python (which should be easy to read):
```
def check_assigns(a):
  for giver,givee in a:
    if giver == givee or giver in block and block[giver] == givee:
      return False
  return True

def match(givers):
  recipients = list(givers)
  while True:
    random.shuffle(recipients)
    assigns = list(map(lambda g,r: (g,r),givers,recipients))
    if check_assigns(assigns): return assigns
```

**Reply by Pr0mises** [reply]
For this blocklist:
```
[block]
susan: paul
paul: susan
mary: kent
kent: mary
esther: dave
dave: esther
```
It could be useful to have an edit field on each user’s profile with a dropdown to manually exclude specific people from the Secret Santa draw.
The admin could also set these exclusions globally for all users before running the Secret Santa.
I'm not yet sure where exactly this exclusion option would best fit in the UI, though.
